plugins/spectate/spectate.cpp

```diff
@@ -36,51 +47,323 @@ using namespace DFHack;
+            config.animals = pconfig.ival(ANIMALS);
+            config.hostiles = pconfig.ival(HOSTILES);
+            config.visitors = pconfig.ival(VISITORS);
+            pause_lock->unlock();
+            SetUnpauseState(config.unpause);
+        }
+    }
+
+    bool FollowADwarf() {
+        if (enabled && !World::ReadPauseState()) {
+            df::coord viewMin = Gui::getViewportPos();
+            df::coord viewMax{viewMin};
+            const auto &dims = Gui::getDwarfmodeViewDims().map().second;
+            viewMax.x += dims.x - 1;
+            viewMax.y += dims.y - 1;
+            viewMax.z = viewMin.z;
+            std::vector<df::unit*> units;
+            static auto add_if = [&](std::function<bool(df::unit*)> check) {
+                for (auto unit : world->units.active) {
+                    if (check(unit)) {
+                        units.push_back(unit);
+                    }
+                }
+            };
+            static auto valid = [](df::unit* unit) {
+                if (Units::isAnimal(unit)) {
+                    return config.animals;
+                }
+                if (Units::isVisiting(unit)) {
+                    return config.visitors;
+                }
+                if (Units::isDanger(unit)) {
+                    return config.hostiles;
+                }
+                return true;
+            };
+            static auto calc_extra_weight = [](size_t idx, double r1, double r2) {
+                switch(idx) {
+                    case 0:
+                        return r2;
+                    case 1:
+                        return (r2-r1)/1.3;
+                    case 2:
+                        return (r2-r1)/2;
+                    default:
+                        return 0.0;
+                }
+            };
+            /// Collecting our choice pool
+            ///////////////////////////////
+            std::array<int32_t, 10> ranges{};
+            std::array<bool, 5> range_exists{};
+            static auto build_range = [&](size_t idx){
+                size_t first = idx * 2;
+                size_t second = idx * 2 + 1;
+                size_t previous = first - 1;
+                // first we get the end of the range
+                ranges[second] = units.size() - 1;
+                // then we calculate whether this indicates there is a range or not
+                if (first != 0) {
+                    range_exists[idx] = ranges[second] > ranges[previous];
+                } else {
+                    range_exists[idx] = ranges[second] >= 0;
+                }
+                // lastly we set the start of the range
+                ranges[first] = ranges[previous] + (range_exists[idx] ? 1 : 0);
+            };
+
+            /// RANGE 0 (in view + working)
+            // grab valid working units
+            add_if([](df::unit* unit) {
+                return valid(unit) && Units::isCitizen(unit, true) && unit->job.current_job;
```

Comment:
```suggestion
                return valid(unit) && unit->job.current_job;
```
Perhaps for in view "working" we can dis-regard whether they are a citizen or not. They probably are.. But if not that might be interesting.